### PR TITLE
Factorize the file filtering of style-analyze into lookout-sdk-ml

### PR DESCRIPTION
**Pull request by warenlg**
Signed-off-by: Waren Long <waren@sourced.tech>

About https://github.com/src-d/style-analyzer/issues/502

### CODE REVIEW

File: lookout/core/lib.py [hunk 1]
```diff
@@ -86,21 +91,57 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+        except NonUTF8ContentException:
+            # skip files that can't be parsed because of UTF-8 decoding errors.
+            continue
+        if res.status == 0 and res.language.lower() == language.lower():
+            uast = res.uast
+            path = filename
+            n_parsed += 1
+            with open(filename) as f:
+                content = f.read().encode("utf-8")
+            if len(max(content.splitlines(), key=len, default=b"")) <= line_length_limit:
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]: How about

```python
with open(filename, "rb") as f:
    content = f.read()
```

Thus you don't have to convert to bytes.

Comment (warenlg) [reply lookout/core/lib.py]: yes indeed

File: lookout/core/lib.py [hunk 2]
```diff
@@ -86,21 +91,57 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+        try:
+            res = client.parse(filename)
+        except NonUTF8ContentException:
+            # skip files that can't be parsed because of UTF-8 decoding errors.
+            continue
+        if res.status == 0 and res.language.lower() == language.lower():
+            uast = res.uast
+            path = filename
+            n_parsed += 1
+            with open(filename, "rb") as f:
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]: We need to filter by line length before parsing with Babelfish

File: lookout/core/lib.py [hunk 3]
```diff
@@ -86,21 +91,57 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+    :param language: Language to consider. Will discard the other languages.
     :param log: logger to use to report the number of excluded files.
     :return: files passed through the filter and the number of files which were excluded.
     """
     passed = []
-    for file in files.values():
-        if len(max(file.content.splitlines(), key=len, default=b"")) <= line_length_limit:
-            passed.append(file)
+    n_parsed = 0
+    for filename in tqdm(filter_filepaths(filenames)):
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]: We cannot do this progress tracking. This is a library, it does not care about the details.

Add another argument which would be `lambda x: x` by default and replace `tqdm`.

File: lookout/core/lib.py [hunk 4]
```diff
@@ -86,21 +91,57 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+    exclude_compiled_pattern = re.compile(exclude_pattern) if exclude_pattern else None
+    for filepath in filepaths:
+        if not isfile(filepath):
+            continue
+        if exclude_compiled_pattern and exclude_compiled_pattern.search(filepath):
+            continue
+        yield filepath
+
+
+def filter_files(filenames: Iterable[str], line_length_limit: int, client: BblfshClient,
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]: Since we do parsing here, this should be renamed to `parse_files`

File: requirements.txt [hunk 5]
```diff
@@ -9,3 +9,4 @@ psycopg2-binary==2.7.5
 modelforge==0.9.3
 bblfsh==2.12.6
 lookout-sdk==0.2.0
+tqdm==4.28.1
```
Comment (vmarkovtsev) [inline requirements.txt]: So this and the other mention of `tqdm` are to be removed

File: lookout/core/lib.py [hunk 6]
```diff
@@ -86,21 +91,57 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+                 language: str, log: logging.Logger = None) -> Iterable[File]:
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]: We return `Iterator`, not `Iterable`.

cc/ @m09 - this is what I was talking about.

Comment (m09) [reply lookout/core/lib.py]: Yes I'll create an issue because we have this problem all over the codebase

**Comment by warenlg** [comment]
Now, it looks better to me, at least I understand what's going on during filtering
```
DEBUG:quality_report_noisy:excluded 52/302 files based on their path
DEBUG:quality_report_noisy:excluded 31/250 javascript files by max line length 150
DEBUG:quality_report_noisy:excluded 62/219 files due to parsing issues
DEBUG:quality_report_noisy:excluded 0/157 javascript files by max overall size 5242880
```

File: lookout/core/lib.py [hunk 7]
```diff
@@ -87,38 +91,83 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
     :param log: logger to use to report the number of excluded files.
     :return: files passed through the filter and the number of files which were excluded.
     """
+    n_parsed = 0
     line_passed = []
-    for key, file in files.items():
-        if len(max(file.content.splitlines(), key=len, default=b"")) <= line_length_limit:
-            line_passed.append(key)
+    n_filenames_filtered = len(list(filter_filepaths(filenames)))
+    for filename in progress_tracker(filter_filepaths(filenames)):
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]: We don't want to parse everything just to realize that we drop some fraction due to `overall_size_limit`

File: lookout/core/lib.py [hunk 8]
```diff
@@ -87,38 +91,104 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+        passed.append(filepath)
+    return passed
+
+
+def filter_files_by_line_length(filepaths: Iterable[str], line_length_limit: int) -> Iterable[str]:
+    """
+    Filter files that have lines longer than `line_length_limit`.
+
+    :param filepaths: paths to the files to filter.
+    :param line_length_limit: maximum line length to accept a file.
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]:
```suggestion
    :param line_length_limit: Maximum line length to accept a file. \
                              We measure the length in bytes, not in Unicode characters.
```

File: lookout/core/lib.py [hunk 9]
```diff
@@ -87,38 +91,104 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+    :return: files passed through the line-length filter.
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]:
```suggestion
    :return: files passed through the maximum line length filter.
```

File: lookout/core/lib.py [hunk 10]
```diff
@@ -87,38 +91,104 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+                ) -> Iterable[File]:
+    """
+    Parse files with Babelfish.
+
+    If a file has lines longer than `line_length_limit`, it is skipped. If the summed size of \
+    parsed files exceeds `overall_size_limit` the rest of the files is skipped. Files paths are \
+    filtered with `filter_filepaths()`. The order in which the files are parsed is random - \
+    and hence different from `filepaths`.
+
+    :param filepaths: paths to the files to filter.
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]:
```suggestion
    :param filepaths: File paths to filter.
```

File: lookout/core/lib.py [hunk 11]
```diff
@@ -87,38 +91,104 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
     :param line_length_limit: maximum line length to accept a file.
     :param overall_size_limit: maximum cumulative files size in bytes. \
                                The files are discarded after reaching this limit.
+    :param client: Babelfish client. Babelfish server should be started accordingly.
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]:
```suggestion
    :param client: Babelfish client instance. The Babelfish server should be running.
```

File: lookout/core/lib.py [hunk 12]
```diff
@@ -87,38 +91,104 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
-    if log is not None:
-        log.debug("excluded %d/%d files by max line length %d",
-                  len(files) - len(line_passed), len(files), line_length_limit)
-    line_passed.sort()
     random.seed(random_state)
-    line_passed = random.sample(line_passed, k=len(line_passed))
-    size = 0
+    filepaths_filtered = filter_filepaths(filepaths)
+    filepaths_filtered = random.sample(filepaths_filtered, k=len(filepaths_filtered))
+    files_filtered_by_line_length = filter_files_by_line_length(filepaths_filtered,
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]: You should swap here: first filter by line length and then shuffle.

Comment (vmarkovtsev) [reply lookout/core/lib.py]: How about renaming `filter_filepaths` to `filter_files_by_path`? The two functions will be uniformly named then.

File: lookout/core/lib.py [hunk 13]
```diff
@@ -87,38 +91,104 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+                                    language=res.language.lower()))
     if log is not None:
-        log.debug("excluded %d/%d files by max overall size %d",
-                  len(line_passed) - len(size_passed), len(line_passed), overall_size_limit)
+        log.debug("excluded %d/%d files based on their path",
+                  len(filepaths) - len(filepaths_filtered), len(filepaths))
+        log.debug("excluded %d/%d %s files by max line length %d",
+                  len(filepaths_filtered) - len(files_filtered_by_line_length),
+                  len(filepaths_filtered), language, line_length_limit)
+        log.debug("excluded %d/%d %s files due to parsing issues",
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]:
```suggestion
        log.debug("excluded %d/%d %s files due to parsing problems",
```

File: lookout/core/tests/test_lib.py [hunk 14]
```diff
@@ -80,23 +81,30 @@ def test_files_by_language(self):
...
+            try:
+                bblfsh_client = BblfshClient("0.0.0.0:9432")
+                filtered = parse_files(filepaths=[tmp1.name, tmp2.name], line_length_limit=80,
+                                       overall_size_limit=5 << 20, client=bblfsh_client,
+                                       language="javascript", log=Log())
+                self.assertEqual(len(filtered), 1)
+                self.assertEqual(filtered[0].content, b"hello")
+                self.assertTrue(logged)
+            finally:
+                bblfsh_client._channel.close()
```
Comment (vmarkovtsev) [inline lookout/core/tests/test_lib.py]: Can you please add a test for filtering by max line length? You can take samples from the test files in `dirname(__file__)`

Comment (warenlg) [reply lookout/core/tests/test_lib.py]: Isn't that a bad practice in general as those test files can be modified later ?

File: lookout/core/lib.py [hunk 15]
```diff
@@ -87,38 +91,107 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+def filter_files_by_path(filepaths: Iterable[str], exclude_pattern: Optional[str] = None,
+                         ) -> Iterable[str]:
     """
-    Filter files based on their maximum line length.
+    Filter out files by specific patterns in their path.
+
+    :param filepaths: Iterable of file paths to examine.
+    :param exclude_pattern: Regular expression to search in file paths. The matched files \
+                            are excluded from the result. If it is None, we use the "garbage" \
+                            pattern defined in lookout.core.langs. We use "" to not filter \
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]: If it is an empty string, filtering is disabled.

File: lookout/core/lib.py [hunk 16]
```diff
@@ -87,38 +91,107 @@ def files_by_language(files: Iterable[File]) -> Dict[str, Dict[str, File]]:
...
+    :param filepaths: File paths to filter.
+    :param line_length_limit: Maximum line length to accept a file.
+    :param overall_size_limit: Maximum cumulative files size in bytes. \
+                               The files are discarded after reaching this limit.
+    :param client: Babelfish client instance. The Babelfish server should be running.
+    :param language: Language to consider. Will discard the other languages.
+    :param random_state: Random generator state for shuffling the files.
+    :param progress_tracker: Optional progress metric whenn iterating over the input files.
+    :param log: Logger to use to report the number of excluded files.
+    :return: Files passed through the filter and the number of files which were excluded.
```
Comment (vmarkovtsev) [inline lookout/core/lib.py]:
```suggestion
    :return: `File`-s with parsed UASTs and which passed through the filters.
```

File: lookout/core/tests/test_lib.py (hunk 14 above)
Comment (vmarkovtsev) [reply lookout/core/tests/test_lib.py]: Yeah, not a rock solid solution, but I think the way you coded it is fine

**Comment by warenlg** [comment]
@vmarkovtsev Should be fine now

**Pull request merged**